### PR TITLE
Zoom 3

### PR DESCRIPTION
**Pull request by sarveshpro**
Changes in this pull request:
- Refactored map component
- Time Travel Feature (Former Zoom 3)
- Vegetation Change on some projects
- multiple satellites sources, select range (2014-2021)
- custom node API integration
- removed Mapbox token dependency

Requires `SITE_IMAGERY_API_URL` in env

Platforms in use for this feature - 
- Google Earth Engine
- Planet Labs

dragging and panning on time travel is still buggy, but it works. please test the entire projects map flow. Home -> Single Project -> Time Travel and back

### CODE REVIEW

**Comment by norbertschuler** [comment]
- [ ] @sarveshpro please run `npm run lint:errors -- --fix` and fix the remaining lint errors (which could indicate possible issues).

File: src/features/projects/components/maps/Explore.tsx [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+  }, [exploreExpanded]);
+
+  return (
+    <>
+      {exploreForests ? (
+        <Source
+          id="forests"
+          type="raster"
+          tiles={[
+            'https://tiles.arcgis.com/tiles/lKUTwQ0dhJzktt4g/arcgis/rest/services/Forest_Denisty_V2/MapServer/tile/{z}/{y}/{x}',
```
Comment (norbertschuler) [inline src/features/projects/components/maps/Explore.tsx:212]: Can we be sure this URL does not change. Should be extract it into an environment variable?

File: src/features/projects/components/maps/Explore.tsx [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+            })}
+        </LayerManager>
+      ) : null}
+
+      {explorePotential ? (
+        <Source
+          id="potential"
+          type="raster"
+          tiles={[
+            'https://tiles.arcgis.com/tiles/lKUTwQ0dhJzktt4g/arcgis/rest/services/Restoration_Potential_Bastin_2019_V3/MapServer/tile/{z}/{y}/{x}',
```
Comment (norbertschuler) [inline src/features/projects/components/maps/Explore.tsx:256]: same as above

File: src/features/projects/components/maps/SatelliteLayer.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+export default function SatelliteLayer({ }: Props): ReactElement {
+    return (
+        <>
+            <Source
+                id="satellite"
+                type="raster"
+                tiles={[
+                    'https://services.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
```
Comment (norbertschuler) [inline src/features/projects/components/maps/SatelliteLayer.tsx:15]: Can we be sure this URL does not change. Should be extract it into an environment variable?

File: src/utils/maps/getMapStyle.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import defaultStyle from '../../../public/data/styles/root.json';
+import openStreetMap from '../../../public/data/styles/OpenStreetMap.json';
+
+export default function getMapStyle(style: any) {
+  return new Promise<any>(async (resolve, reject) => {
+    if (style === 'default') {
+      let result = await fetchTiles(
+        defaultStyle,
+        'https://basemaps.arcgis.com/arcgis/rest/services/World_Basemap_v2/VectorTileServer'
```
Comment (norbertschuler) [inline src/utils/maps/getMapStyle.ts:9]: Can we be sure this URL does not change. Should be extract it into an environment variable?

File: src/utils/maps/getMapStyle.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      );
+      resolve(result);
+    } else if (style === 'openStreetMap') {
+      let result = await fetchTiles(
+        openStreetMap,
+        'https://basemaps.arcgis.com/arcgis/rest/services/OpenStreetMap_v2/VectorTileServer'
```
Comment (norbertschuler) [inline src/utils/maps/getMapStyle.ts:15]: same of above

File: .env.local.example [hunk 6]
```diff
@@ -15,3 +15,5 @@ SENTRY_ORG=plant-for-the-planet
 SENTRY_PROJECT=app-20-next
 SENTRY_AUTH_TOKEN=
 VERCEL_GITHUB_COMMIT_SHA=
+#for vegetation change
+GEE_API_URL=http://localhost:8000
```
[the remark this answers is not in the document]
Comment (norbertschuler) [reply .env.local.example:19]: I found https://raster.plant-for-the-planet.org - but probably we do not have anything setup for develop or staging, do we?

**Comment by norbertschuler** [comment]
- [ ] When you switch of projects the markers on the map are not hidden any more compare to the current production website. Is this wanted or a bug?

Bildschirmfoto 2021-03-03 um 15 50 14Bildschirmfoto 2021-03-03 um 15 51 01

File: .env.local.example (hunk 6 above)
[the remark this answers is not in the document]
Comment (sagararyal) [reply .env.local.example:19]: maybe change name to `SITE_IMAGERY_API_URL`

File: next.config.js [hunk 7]
```diff
@@ -42,161 +43,152 @@ const nextauthUrl = process.env.NEXTAUTH_URL
...
+    }
+    return config;
+  },
+  basePath,
+  // your config for other plugins or the general next.js here...
+  devIndicators: {
+    autoPrerender: false,
+  },
+  env: {
+    MAPBOXGL_ACCESS_TOKEN: process.env.MAPBOXGL_ACCESS_TOKEN,
```
Comment (sagararyal) [inline next.config.js:110]: do we still use Mapbox?

File: pages/api/landsat-imagery.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
+import ee from '@google/earthengine';
+import Cors from 'cors';
+
+var privateKey = {
+  private_key: process.env.GCP_PRIVATE_KEY,
+  client_email: process.env.GCP_SERVICE_ACCOUNT,
```
[the remark this answers is not in the document]
Comment (sagararyal) [reply pages/api/landsat-imagery.js:6]: I don't think this is necessary here, or is it? Don't we use GEE on the Node server?

File: public/data/planet/aoi.json [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{
+        "type": "Polygon",
```
Comment (sagararyal) [inline public/data/planet/aoi.json:2]: @sarveshpro is this a test file? why do we need this?

File: src/features/projects/components/maps/Explore.tsx (hunk 2 above)
Comment (sagararyal) [reply src/features/projects/components/maps/Explore.tsx:256]: these should be pretty stable.

File: src/features/projects/components/maps/backup/CompareMaps.tsx [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+
+  React.useEffect(() => {
+    if (before && after) {
+      try {
+        nicfi_data.map((year: any) => {
+          if (year.year === selectedYear1) {
+            if (!before.getSource(`before-imagery-${year.year}`)) {
+              before.addSource(`before-imagery-${year.year}`, {
+                type: 'raster',
+                tiles: [`https://tile-s1.plant-for-the-planet.org/basemaps/v1/planet-tiles/${year.raster}/gmap/{z}/{x}/{y}.png`],
```
[the remark this answers is not in the document]
Comment (sagararyal) [reply src/features/projects/components/maps/backup/CompareMaps.tsx:164]: @sarveshpro weren't the comparison urls coming from node server?

**Review by sagararyal — COMMENTED** [review]
Mostly looks good.

File: package.json [hunk 11]
```diff
@@ -33,7 +33,9 @@
   "dependencies": {
     "@auth0/auth0-react": "^1.2.0",
     "@date-io/date-fns": "1.3.13",
+    "@google/earthengine": "^0.1.243",
```
[the remark this answers is not in the document]
Comment (sarveshpro) [reply package.json:36]: Sorry, this is not needed now. will remove.

File: .env.local.example (hunk 6 above)
[the remark this answers is not in the document]
Comment (sarveshpro) [reply .env.local.example:19]: Sure, we don't have development and staging.

File: pages/api/landsat-imagery.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (sarveshpro) [reply pages/api/landsat-imagery.js:6]: need to remove this file

File: pages/api/landsat-imagery.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        return reject(result);
+      }
+
+      return resolve(result);
+    });
+  });
+}
+
+export default async (req, res) => {
+  // await runMiddleware(req, res, cors);
```
[the remark this answers is not in the document]
Comment (sarveshpro) [reply pages/api/landsat-imagery.js:29]: not being used anymore

File: src/features/projects/components/maps/Explore.tsx (hunk 1 above)
Comment (sarveshpro) [reply src/features/projects/components/maps/Explore.tsx:212]: This is hosted in our esri account so not unless someone decides to delete it. and if it changes it won't crash the app.

File: next.config.js (hunk 7 above)
Comment (sarveshpro) [reply next.config.js:110]: yes in one place, will remove in this branch

File: public/data/planet/aoi.json (hunk 9 above)
Comment (sarveshpro) [reply public/data/planet/aoi.json:2]: This is the available area of Planet Labs data. I check if the project is inside it or fallback to sentinel

File: src/features/projects/components/maps/ProjectPolygon.tsx [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    id: string | undefined;
+    geoJson: Object | null;
+}
+
+export default function ProjectPolygon({ id, geoJson }: Props): ReactElement {
+    return (
+        <>
+            <Source id={id ? id : "singleProject"} type="geojson" data={geoJson}>
+                <Layer
+                    id={id ? id : "ploygonOutline"}
```
[the remark this answers is not in the document]
Comment (sarveshpro) [reply src/features/projects/components/maps/ProjectPolygon.tsx:14]: typo, fixed.

**Comment by norbertschuler** [comment]
- [ ] We still have GEE_API_URL, GCP_PRIVATE_KEY and GCP_SERVICE_ACCOUNT configured in the Vercel deployments while SITE_IMAGERY_API_URL is missing. Can we clean this up and add the new missing env variable?

**Comment by norbertschuler** [comment]
- [ ] Is https://dashboard.heroku.com/apps/planet-geo-api the production version and https://dashboard.heroku.com/apps/gee-preview for development? Maybe we can group them into a pipe?

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (norbertschuler) [reply package.json:36]: @sarveshpro but you have not removed it?

File: src/features/donations/screens/TreeDonation.tsx [hunk 14]
```diff
@@ -93,6 +93,7 @@ function TreeDonation({
     }
 
     setMinAmt(getMinimumAmountForCurrency(currency));
+    console.log(minAmt);
```
Comment (norbertschuler) [inline src/features/donations/screens/TreeDonation.tsx:96]: Please remove unnecessary console.log statements again once this PR is ready to get merged.

**Review by norbertschuler — COMMENTED** [review]
I added `SITE_IMAGERY_API_URL` to all env files as well as app.json.

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (norbertschuler) [reply package.json]: I removed it myself now as I wanted to trigger a rebuild anymore to see if the env variables of Vercel work now.

**Review by norbertschuler — APPROVED** [review]
LGFM and as `SITE_IMAGERY_API_URL` once was not available on Vercel I saw without that env variable the new features just do not appear. So I would approve this PR to be merged.

File: src/features/donations/screens/TreeDonation.tsx (hunk 14 above)
Comment (sarveshpro) [reply src/features/donations/screens/TreeDonation.tsx]: removed

**Comment by norbertschuler** [comment]
I tested this PR with Chrome/Firefox/Safari on MacOS, but also with Chrome/Firefox on Win10 and some mobile devices at Browserstack. It looks quite ok and seems to be also compatible with iOS 11 or Android 6 devices (but maybe is getting a little slow on older devices).

File: src/features/common/InputTypes/BootstrapInput.tsx [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        '-apple-system',
+        'BlinkMacSystemFont',
+        '"Segoe UI"',
+        'Roboto',
+        '"Helvetica Neue"',
+        'Arial',
+        'sans-serif',
+        '"Apple Color Emoji"',
+        '"Segoe UI Emoji"',
+        '"Segoe UI Symbol"',
```
Comment (sagararyal) [inline src/features/common/InputTypes/BootstrapInput.tsx:29]: these fonts are not our fonts,

Also, I think we had removed Bootstrap some time ago; is there a reason we need to add it again?

File: src/features/user/ManageProjects/components/ProjectSites.tsx [hunk 16]
```diff
@@ -237,7 +260,7 @@ export default function ProjectSites({
                       {...viewport}
                       center={[longitude, latitude]}
                       zoom={[zoom]}
-                      style="mapbox://styles/mapbox/streets-v11?optimize=true" // eslint-disable-line
+                      style={style} // eslint-disable-line
```
Comment (sagararyal) [inline src/features/user/ManageProjects/components/ProjectSites.tsx:263]: We need to update the default style to include road level details.

**Review by sagararyal — COMMENTED** [review]
@sarveshpro

File: src/features/common/InputTypes/BootstrapInput.tsx (hunk 15 above)
Comment (sarveshpro) [reply src/features/common/InputTypes/BootstrapInput.tsx:29]: No this is a custom component, using material UI

Comment (sarveshpro) [reply src/features/common/InputTypes/BootstrapInput.tsx:29]: fonts I'll remove

File: src/features/user/ManageProjects/components/ProjectSites.tsx (hunk 16 above)
Comment (sarveshpro) [reply src/features/user/ManageProjects/components/ProjectSites.tsx:263]: here I think we are using OSM street map, it should have road level data

File: src/features/common/InputTypes/BootstrapInput.tsx (hunk 15 above)
Comment (sarveshpro) [reply src/features/common/InputTypes/BootstrapInput.tsx:29]: Done

**Pull request merged**